File: android/app/src/main/java/com/irrgenius/android/data/import/CSVImportService.kt
package com.irrgenius.android.data.import

import com.irrgenius.android.data.models.*
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.*

/**
 * Service for importing calculation data from CSV files
 */
class CSVImportService {
    
    companion object {
        private val SUPPORTED_DELIMITERS = listOf(',', ';', '\t', '|')
        private val DATE_FORMATTERS = listOf(
            DateTimeFormatter.ofPattern("yyyy-MM-dd"),
            DateTimeFormatter.ofPattern("MM/dd/yyyy"),
            DateTimeFormatter.ofPattern("dd/MM/yyyy"),
            DateTimeFormatter.ofPattern("yyyy/MM/dd")
        )
    }
    
    /**
     * Imports CSV data from an InputStream
     */
    suspend fun importCSV(
        inputStream: InputStream,
        delimiter: Char = ',',
        hasHeaders: Boolean = true
    ): ImportResult {
        return try {
            val reader = BufferedReader(InputStreamReader(inputStream))
            val lines = reader.readLines()
            reader.close()
            
            if (lines.isEmpty()) {
                throw ImportException("File is empty")
            }
            
            val actualDelimiter = detectDelimiter(lines.first(), delimiter)
            val parsedData = parseCSVLines(lines, actualDelimiter, hasHeaders)
            
            ImportResult(
                headers = parsedData.headers,
                rows = parsedData.rows,
                detectedFormat = ImportFormat.CSV(actualDelimiter, hasHeaders),
                suggestedMapping = suggestColumnMapping(parsedData.headers),
                validationErrors = emptyList()
            )
        } catch (e: Exception) {
            throw ImportException("Failed to parse CSV: ${e.message}", e)
        }
    }
    
    /**
     * Validates imported data and converts to SavedCalculation objects
     */
    suspend fun validateAndConvert(
        importResult: ImportResult,
        columnMapping: Map<String, CalculationField>,
        calculationType: CalculationMode,
        projectId: String? = null
    ): ValidationResult {
        val validationErrors = mutableListOf<ValidationError>()
        val validCalculations = mutableListOf<SavedCalculation>()
        
        importResult.rows.forEachIndexed { rowIndex, row ->
            try {
                val calculation = convertRowToCalculation(
                    row = row,
                    headers = importResult.headers,
                    columnMapping = columnMapping,
                    calculationType = calculationType,
                    projectId = projectId,
                    rowIndex = rowIndex
                )
                
                // Validate the calculation
                calculation.validate()
                validCalculations.add(calculation)
                
            } catch (e: Exception) {
                validationErrors.add(
                    ValidationError(
                        row = rowIndex + 1,
                        column = null,
                        message = e.message ?: "Unknown validation error",
                        severity = ValidationSeverity.ERROR
                    )
                )
            }
        }
        
        return ValidationResult(
            validCalculations = validCalculations,
            validationErrors = validationErrors,
            totalRows = importResult.rows.size,
            validRows = validCalculations.size
        )
    }
    
    /**
     * Detects the most likely delimiter in a CSV line
     */
    private fun detectDelimiter(line: String, preferredDelimiter: Char): Char {
        if (line.contains(preferredDelimiter)) {
            return preferredDelimiter
        }
        
        return SUPPORTED_DELIMITERS.maxByOrNull { delimiter ->
            line.count { it == delimiter }
        } ?: ','
    }
    
    /**
     * Parses CSV lines into headers and rows
     */
    private fun parseCSVLines(
        lines: List<String>,
        delimiter: Char,
        hasHeaders: Boolean
    ): ParsedCSVData {
        val headers = if (hasHeaders && lines.isNotEmpty()) {
            parseCSVLine(lines.first(), delimiter)
        } else {
            // Generate default headers if no headers provided
            val firstLine = lines.firstOrNull() ?: return ParsedCSVData(emptyList(), emptyList())
            val columnCount = parseCSVLine(firstLine, delimiter).size
            (1..columnCount).map { "Column $it" }
        }
        
        val dataStartIndex = if (hasHeaders) 1 else 0
        val rows = lines.drop(dataStartIndex).map { line ->
            parseCSVLine(line, delimiter)
        }.filter { it.isNotEmpty() } // Filter out empty rows
        
        return ParsedCSVData(headers, rows)
    }
    
    /**
     * Parses a single CSV line, handling quoted values
     */
    private fun parseCSVLine(line: String, delimiter: Char): List<String> {
        val result = mutableListOf<String>()
        val current = StringBuilder()
        var inQuotes = false
        var i = 0
        
        while (i < line.length) {
            val char = line[i]
            
            when {
                char == '"' -> {
                    if (inQuotes && i + 1 < line.length && line[i + 1] == '"') {
                        // Escaped quote
                        current.append('"')
                        i++ // Skip next quote
                    } else {
                        // Toggle quote state
                        inQuotes = !inQuotes
                    }
                }
                char == delimiter && !inQuotes -> {
                    result.add(current.toString().trim())
                    current.clear()
                }
                else -> {
                    current.append(char)
                }
            }
            i++
        }
        
        result.add(current.toString().trim())
        return result
    }
    
    /**
     * Suggests column mapping based on header names
     */
    private fun suggestColumnMapping(headers: List<String>): Map<String, CalculationField> {
        val mapping = mutableMapOf<String, CalculationField>()
        
        headers.forEach { header ->
            val normalizedHeader = header.lowercase().replace(Regex("[^a-z0-9]"), "")
            
            val suggestedField = when {
                normalizedHeader.contains("initial") || normalizedHeader.contains("investment") -> CalculationField.INITIAL_INVESTMENT
                normalizedHeader.contains("outcome") || normalizedHeader.contains("exit") -> CalculationField.OUTCOME_AMOUNT
                normalizedHeader.contains("time") || normalizedHeader.contains("month") || normalizedHeader.contains("duration") -> CalculationField.TIME_IN_MONTHS
                normalizedHeader.contains("irr") || normalizedHeader.contains("return") -> CalculationField.IRR
                normalizedHeader.contains("name") || normalizedHeader.contains("title") -> CalculationField.NAME
                normalizedHeader.contains("note") || normalizedHeader.contains("comment") -> CalculationField.NOTES
                normalizedHeader.contains("date") || normalizedHeader.contains("created") -> CalculationField.DATE
                normalizedHeader.contains("unit") && normalizedHeader.contains("price") -> CalculationField.UNIT_PRICE
                normalizedHeader.contains("success") && normalizedHeader.contains("rate") -> CalculationField.SUCCESS_RATE
                normalizedHeader.contains("outcome") && normalizedHeader.contains("unit") -> CalculationField.OUTCOME_PER_UNIT
                normalizedHeader.contains("investor") && normalizedHeader.contains("share") -> CalculationField.INVESTOR_SHARE
                normalizedHeader.contains("fee") -> CalculationField.FEE_PERCENTAGE
                else -> null
            }
            
            suggestedField?.let { mapping[header] = it }
        }
        
        return mapping
    }
    
    /**
     * Converts a CSV row to a SavedCalculation object
     */
    private fun convertRowToCalculation(
        row: List<String>,
        headers: List<String>,
        columnMapping: Map<String, CalculationField>,
        calculationType: CalculationMode,
        projectId: String?,
        rowIndex: Int
    ): SavedCalculation {
        val fieldValues = mutableMapOf<CalculationField, String>()
        
        // Map row values to calculation fields
        headers.forEachIndexed { index, header ->
            if (index < row.size) {
                columnMapping[header]?.let { field ->
                    fieldValues[field] = row[index]
                }
            }
        }
        
        // Extract and validate field values
        val name = fieldValues[CalculationField.NAME]?.takeIf { it.isNotBlank() }
            ?: "Imported Calculation ${rowIndex + 1}"
        
        val initialInvestment = fieldValues[CalculationField.INITIAL_INVESTMENT]?.let { 
            parseDouble(it, "Initial Investment", rowIndex + 1)
        }
        
        val outcomeAmount = fieldValues[CalculationField.OUTCOME_AMOUNT]?.let {
            parseDouble(it, "Outcome Amount", rowIndex + 1)
        }
        
        val timeInMonths = fieldValues[CalculationField.TIME_IN_MONTHS]?.let {
            parseDouble(it, "Time in Months", rowIndex + 1)
        }
        
        val irr = fieldValues[CalculationField.IRR]?.let {
            parseDouble(it, "IRR", rowIndex + 1)
        }
        
        val unitPrice = fieldValues[CalculationField.UNIT_PRICE]?.let {
            parseDouble(it, "Unit Price", rowIndex + 1)
        }
        
        val successRate = fieldValues[CalculationField.SUCCESS_RATE]?.let {
            parseDouble(it, "Success Rate", rowIndex + 1)
        }
        
        val outcomePerUnit = fieldValues[CalculationField.OUTCOME_PER_UNIT]?.let {
            parseDouble(it, "Outcome Per Unit", rowIndex + 1)
        }
        
        val investorShare = fieldValues[CalculationField.INVESTOR_SHARE]?.let {
            parseDouble(it, "Investor Share", rowIndex + 1)
        }
        
        val feePercentage = fieldValues[CalculationField.FEE_PERCENTAGE]?.let {
            parseDouble(it, "Fee Percentage", rowIndex + 1)
        }
        
        val notes = fieldValues[CalculationField.NOTES]
        
        val createdDate = fieldValues[CalculationField.DATE]?.let { 
            parseDate(it, rowIndex + 1) 
        } ?: LocalDateTime.now()
        
        return SavedCalculation.createValidated(
            name = name,
            calculationType = calculationType,
            createdDate = createdDate,
            modifiedDate = LocalDateTime.now(),
            projectId = projectId,
            initialInvestment = initialInvestment,
            outcomeAmount = outcomeAmount,
            timeInMonths = timeInMonths,
            irr = irr,
            unitPrice = unitPrice,
            successRate = successRate,
            outcomePerUnit = outcomePerUnit,
            investorShare = investorShare,
            feePercentage = feePercentage,
            notes = notes
        )
    }
    
    /**
     * Parses a string to Double with error handling
     */
    private fun parseDouble(value: String, fieldName: String, rowNumber: Int): Double {
        return try {
            // Remove common currency symbols and formatting
            val cleanValue = value.replace(Regex("[$,€£¥%\\s]"), "")
            cleanValue.toDouble()
        } catch (e: NumberFormatException) {
            throw ImportException("Invalid number format in row $rowNumber for field '$fieldName': '$value'")
        }
    }
    
    /**
     * Parses a string to LocalDateTime with error handling
     */
    private fun parseDate(value: String, rowNumber: Int): LocalDateTime {
        for (formatter in DATE_FORMATTERS) {
            try {
                return LocalDateTime.parse(value, formatter.withTime(java.time.LocalTime.MIDNIGHT))
            } catch (e: DateTimeParseException) {
                // Try next formatter
            }
        }
        throw ImportException("Invalid date format in row $rowNumber: '$value'")
    }
    
    private fun DateTimeFormatter.withTime(time: java.time.LocalTime): DateTimeFormatter {
        return DateTimeFormatter.ofPattern("${this.toString()} HH:mm:ss")
    }
}

/**
 * Data class for parsed CSV data
 */
private data class ParsedCSVData(
    val headers: List<String>,
    val rows: List<List<String>>
)

/**
 * Enum representing calculation fields that can be imported
 */
enum class CalculationField(val displayName: String) {
    NAME("Calculation Name"),
    INITIAL_INVESTMENT("Initial Investment"),
    OUTCOME_AMOUNT("Outcome Amount"),
    TIME_IN_MONTHS("Time in Months"),
    IRR("IRR (%)"),
    UNIT_PRICE("Unit Price"),
    SUCCESS_RATE("Success Rate (%)"),
    OUTCOME_PER_UNIT("Outcome Per Unit"),
    INVESTOR_SHARE("Investor Share (%)"),
    FEE_PERCENTAGE("Fee Percentage (%)"),
    NOTES("Notes"),
    DATE("Date")
}

/**
 * Import format specification
 */
sealed class ImportFormat {
    data class CSV(val delimiter: Char, val hasHeaders: Boolean) : ImportFormat()
    data class Excel(val sheetName: String, val hasHeaders: Boolean) : ImportFormat()
}

/**
 * Result of CSV import operation
 */
data class ImportResult(
    val headers: List<String>,
    val rows: List<List<String>>,
    val detectedFormat: ImportFormat,
    val suggestedMapping: Map<String, CalculationField>,
    val validationErrors: List<ValidationError>
)

/**
 * Result of data validation
 */
data class ValidationResult(
    val validCalculations: List<SavedCalculation>,
    val validationErrors: List<ValidationError>,
    val totalRows: Int,
    val validRows: Int
) {
    val hasErrors: Boolean get() = validationErrors.isNotEmpty()
    val successRate: Double get() = if (totalRows > 0) validRows.toDouble() / totalRows else 0.0
}

/**
 * Validation error information
 */
data class ValidationError(
    val row: Int,
    val column: String?,
    val message: String,
    val severity: ValidationSeverity
)

/**
 * Severity levels for validation errors
 */
enum class ValidationSeverity {
    WARNING,
    ERROR
}

/**
 * Exception thrown during import operations
 */
class ImportException(message: String, cause: Throwable? = null) : Exception(message, cause)

File: ios/IRR Genius/Services/CSVImportService.swift
//
//  CSVImportService.swift
//  IRR Genius
//

import Foundation

/// Service for importing calculation data from CSV files
class CSVImportService {
    
    // MARK: - Constants
    
    private static let supportedDelimiters: [Character] = [",", ";", "\t", "|"]
    private static let dateFormatters: [DateFormatter] = {
        let formatters = [
            "yyyy-MM-dd",
            "MM/dd/yyyy",
            "dd/MM/yyyy",
            "yyyy/MM/dd"
        ].map { pattern in
            let formatter = DateFormatter()
            formatter.dateFormat = pattern
            formatter.locale = Locale(identifier: "en_US_POSIX")
            return formatter
        }
        return formatters
    }()
    
    // MARK: - Public Methods
    
    /// Imports CSV data from a URL
    func importCSV(from url: URL, delimiter: Character = ",", hasHeaders: Bool = true) async throws -> ImportResult {
        guard url.startAccessingSecurityScopedResource() else {
            throw ImportError.fileAccessDenied
        }
        defer { url.stopAccessingSecurityScopedResource() }
        
        do {
            let content = try String(contentsOf: url, encoding: .utf8)
            let lines = content.components(separatedBy: .newlines).filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            
            guard !lines.isEmpty else {
                throw ImportError.emptyFile
            }
            
            let actualDelimiter = detectDelimiter(in: lines.first!, preferred: delimiter)
            let parsedData = parseCSVLines(lines, delimiter: actualDelimiter, hasHeaders: hasHeaders)
            
            return ImportResult(
                headers: parsedData.headers,
                rows: parsedData.rows,
                detectedFormat: .csv(delimiter: actualDelimiter, hasHeaders: hasHeaders),
                suggestedMapping: suggestColumnMapping(for: parsedData.headers),
                validationErrors: []
            )
        } catch {
            if error is ImportError {
                throw error
            } else {
                throw ImportError.parseError("Failed to parse CSV: \(error.localizedDescription)")
            }
        }
    }
    
    /// Validates imported data and converts to SavedCalculation objects
    func validateAndConvert(
        importResult: ImportResult,
        columnMapping: [String: CalculationField],
        calculationType: CalculationMode,
        projectId: UUID? = nil
    ) async throws -> ValidationResult {
        var validationErrors: [ValidationError] = []
        var validCalculations: [SavedCalculation] = []
        
        for (rowIndex, row) in importResult.rows.enumerated() {
            do {
                let calculation = try convertRowToCalculation(
                    row: row,
                    headers: importResult.headers,
                    columnMapping: columnMapping,
                    calculationType: calculationType,
                    projectId: projectId,
                    rowIndex: rowIndex
                )
                
                // Validate the calculation
                try calculation.validate()
                validCalculations.append(calculation)
                
            } catch {
                validationErrors.append(
                    ValidationError(
                        row: rowIndex + 1,
                        column: nil,
                        message: error.localizedDescription,
                        severity: .error
                    )
                )
            }
        }
        
        return ValidationResult(
            validCalculations: validCalculations,
            validationErrors: validationErrors,
            totalRows: importResult.rows.count,
            validRows: validCalculations.count
        )
    }
    
    // MARK: - Private Methods
    
    /// Detects the most likely delimiter in a CSV line
    private func detectDelimiter(in line: String, preferred: Character) -> Character {
        if line.contains(preferred) {
            return preferred
        }
        
        return Self.supportedDelimiters.max { delimiter1, delimiter2 in
            line.filter { $0 == delimiter1 }.count < line.filter { $0 == delimiter2 }.count
        } ?? ","
    }
    
    /// Parses CSV lines into headers and rows
    private func parseCSVLines(_ lines: [String], delimiter: Character, hasHeaders: Bool) -> ParsedCSVData {
        let headers: [String]
        if hasHeaders && !lines.isEmpty {
            headers = parseCSVLine(lines.first!, delimiter: delimiter)
        } else {
            // Generate default headers if no headers provided
            let firstLine = lines.first ?? ""
            let columnCount = parseCSVLine(firstLine, delimiter: delimiter).count
            headers = (1...columnCount).map { "Column \($0)" }
        }
        
        let dataStartIndex = hasHeaders ? 1 : 0
        let rows = lines.dropFirst(dataStartIndex).compactMap { line in
            let parsedLine = parseCSVLine(line, delimiter: delimiter)
            return parsedLine.isEmpty ? nil : parsedLine
        }
        
        return ParsedCSVData(headers: headers, rows: Array(rows))
    }
    
    /// Parses a single CSV line, handling quoted values
    private func parseCSVLine(_ line: String, delimiter: Character) -> [String] {
        var result: [String] = []
        var current = ""
        var inQuotes = false
        var i = line.startIndex
        
        while i < line.endIndex {
            let char = line[i]
            
            switch char {
            case "\"":
                if inQuotes && line.index(after: i) < line.endIndex && line[line.index(after: i)] == "\"" {
                    // Escaped quote
                    current.append("\"")
                    i = line.index(after: i) // Skip next quote
                } else {
                    // Toggle quote state
                    inQuotes.toggle()
                }
            case delimiter where !inQuotes:
                result.append(current.trimmingCharacters(in: .whitespacesAndNewlines))
                current = ""
            default:
                current.append(char)
            }
            
            i = line.index(after: i)
        }
        
        result.append(current.trimmingCharacters(in: .whitespacesAndNewlines))
        return result
    }
    
    /// Suggests column mapping based on header names
    private func suggestColumnMapping(for headers: [String]) -> [String: CalculationField] {
        var mapping: [String: CalculationField] = [:]
        
        for header in headers {
            let normalizedHeader = header.lowercased().replacingOccurrences(of: "[^a-z0-9]", with: "", options: .regularExpression)
            
            let suggestedField: CalculationField?
            switch normalizedHeader {
            case let h where h.contains("initial") || h.contains("investment"):
                suggestedField = .initialInvestment
            case let h where h.contains("outcome") || h.contains("exit"):
                suggestedField = .outcomeAmount
            case let h where h.contains("time") || h.contains("month") || h.contains("duration"):
                suggestedField = .timeInMonths
            case let h where h.contains("irr") || h.contains("return"):
                suggestedField = .irr
            case let h where h.contains("name") || h.contains("title"):
                suggestedField = .name
            case let h where h.contains("note") || h.contains("comment"):
                suggestedField = .notes
            case let h where h.contains("date") || h.contains("created"):
                suggestedField = .date
            case let h where h.contains("unit") && h.contains("price"):
                suggestedField = .unitPrice
            case let h where h.contains("success") && h.contains("rate"):
                suggestedField = .successRate
            case let h where h.contains("outcome") && h.contains("unit"):
                suggestedField = .outcomePerUnit
            case let h where h.contains("investor") && h.contains("share"):
                suggestedField = .investorShare
            case let h where h.contains("fee"):
                suggestedField = .feePercentage
            default:
                suggestedField = nil
            }
            
            if let field = suggestedField {
                mapping[header] = field
            }
        }
        
        return mapping
    }
    
    /// Converts a CSV row to a SavedCalculation object
    private func convertRowToCalculation(
        row: [String],
        headers: [String],
        columnMapping: [String: CalculationField],
        calculationType: CalculationMode,
        projectId: UUID?,
        rowIndex: Int
    ) throws -> SavedCalculation {
        var fieldValues: [CalculationField: String] = [:]
        
        // Map row values to calculation fields
        for (index, header) in headers.enumerated() {
            if index < row.count, let field = columnMapping[header] {
                fieldValues[field] = row[index]
            }
        }
        
        // Extract and validate field values
        let name = fieldValues[.name]?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
            ? fieldValues[.name]!
            : "Imported Calculation \(rowIndex + 1)"
        
        let initialInvestment = try fieldValues[.initialInvestment].flatMap { 
            try parseDouble($0, fieldName: "Initial Investment", rowNumber: rowIndex + 1)
        }
        
        let outcomeAmount = try fieldValues[.outcomeAmount].flatMap {
            try parseDouble($0, fieldName: "Outcome Amount", rowNumber: rowIndex + 1)
        }
        
        let timeInMonths = try fieldValues[.timeInMonths].flatMap {
            try parseDouble($0, fieldName: "Time in Months", rowNumber: rowIndex + 1)
        }
        
        let irr = try fieldValues[.irr].flatMap {
            try parseDouble($0, fieldName: "IRR", rowNumber: rowIndex + 1)
        }
        
        let unitPrice = try fieldValues[.unitPrice].flatMap {
            try parseDouble($0, fieldName: "Unit Price", rowNumber: rowIndex + 1)
        }
        
        let successRate = try fieldValues[.successRate].flatMap {
            try parseDouble($0, fieldName: "Success Rate", rowNumber: rowIndex + 1)
        }
        
        let outcomePerUnit = try fieldValues[.outcomePerUnit].flatMap {
            try parseDouble($0, fieldName: "Outcome Per Unit", rowNumber: rowIndex + 1)
        }
        
        let investorShare = try fieldValues[.investorShare].flatMap {
            try parseDouble($0, fieldName: "Investor Share", rowNumber: rowIndex + 1)
        }
        
        let feePercentage = try fieldValues[.feePercentage].flatMap {
            try parseDouble($0, fieldName: "Fee Percentage", rowNumber: rowIndex + 1)
        }
        
        let notes = fieldValues[.notes]
        
        let createdDate = try fieldValues[.date].flatMap { 
            try parseDate($0, rowNumber: rowIndex + 1) 
        } ?? Date()
        
        return try SavedCalculation(
            name: name,
            calculationType: calculationType,
            createdDate: createdDate,
            modifiedDate: Date(),
            projectId: projectId,
            initialInvestment: initialInvestment,
            outcomeAmount: outcomeAmount,
            timeInMonths: timeInMonths,
            irr: irr,
            unitPrice: unitPrice,
            successRate: successRate,
            outcomePerUnit: outcomePerUnit,
            investorShare: investorShare,
            feePercentage: feePercentage,
            notes: notes
        )
    }
    
    /// Parses a string to Double with error handling
    private func parseDouble(_ value: String, fieldName: String, rowNumber: Int) throws -> Double? {
        let trimmedValue = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedValue.isEmpty else { return nil }
        
        // Remove common currency symbols and formatting
        let cleanValue = trimmedValue.replacingOccurrences(of: "[$,€£¥%\\s]", with: "", options: .regularExpression)
        
        guard let doubleValue = Double(cleanValue) else {
            throw ImportError.invalidNumberFormat(row: rowNumber, field: fieldName, value: value)
        }
        
        return doubleValue
    }
    
    /// Parses a string to Date with error handling
    private func parseDate(_ value: String, rowNumber: Int) throws -> Date? {
        let trimmedValue = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedValue.isEmpty else { return nil }
        
        for formatter in Self.dateFormatters {
            if let date = formatter.date(from: trimmedValue) {
                return date
            }
        }
        
        throw ImportError.invalidDateFormat(row: rowNumber, value: value)
    }
}

// MARK: - Supporting Types

/// Data structure for parsed CSV data
private struct ParsedCSVData {
    let headers: [String]
    let rows: [[String]]
}

/// Enum representing calculation fields that can be imported
enum CalculationField: String, CaseIterable {
    case name = "Calculation Name"
    case initialInvestment = "Initial Investment"
    case outcomeAmount = "Outcome Amount"
    case timeInMonths = "Time in Months"
    case irr = "IRR (%)"
    case unitPrice = "Unit Price"
    case successRate = "Success Rate (%)"
    case outcomePerUnit = "Outcome Per Unit"
    case investorShare = "Investor Share (%)"
    case feePercentage = "Fee Percentage (%)"
    case notes = "Notes"
    case date = "Date"
    
    var displayName: String {
        return self.rawValue
    }
}

/// Import format specification
enum ImportFormat {
    case csv(delimiter: Character, hasHeaders: Bool)
    case excel(sheetName: String, hasHeaders: Bool)
}

/// Result of CSV import operation
struct ImportResult {
    let headers: [String]
    let rows: [[String]]
    let detectedFormat: ImportFormat
    let suggestedMapping: [String: CalculationField]
    let validationErrors: [ValidationError]
}

/// Result of data validation
struct ValidationResult {
    let validCalculations: [SavedCalculation]
    let validationErrors: [ValidationError]
    let totalRows: Int
    let validRows: Int
    
    var hasErrors: Bool { !validationErrors.isEmpty }
    var successRate: Double { totalRows > 0 ? Double(validRows) / Double(totalRows) : 0.0 }
}

/// Validation error information
struct ValidationError {
    let row: Int
    let column: String?
    let message: String
    let severity: ValidationSeverity
}

/// Severity levels for validation errors
enum ValidationSeverity {
    case warning
    case error
}

/// Errors that can occur during import operations
enum ImportError: LocalizedError {
    case fileAccessDenied
    case emptyFile
    case parseError(String)
    case invalidNumberFormat(row: Int, field: String, value: String)
    case invalidDateFormat(row: Int, value: String)
    case unsupportedFormat
    case corruptedFile
    case missingRequiredFields([String])
    
    var errorDescription: String? {
        switch self {
        case .fileAccessDenied:
            return "Unable to access the selected file. Please check file permissions."
        case .emptyFile:
            return "The selected file is empty."
        case .parseError(let message):
            return "Failed to parse file: \(message)"
        case .invalidNumberFormat(let row, let field, let value):
            return "Invalid number format in row \(row) for field '\(field)': '\(value)'"
        case .invalidDateFormat(let row, let value):
            return "Invalid date format in row \(row): '\(value)'"
        case .unsupportedFormat:
            return "File format not supported. Please use CSV or Excel files."
        case .corruptedFile:
            return "File appears to be corrupted or unreadable."
        case .missingRequiredFields(let fields):
            return "Missing required fields: \(fields.joined(separator: ", "))"
        }
    }
}